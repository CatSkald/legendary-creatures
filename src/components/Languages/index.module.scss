.active {
  /* placeholder to be referenced from styles */
}

.languageMenu {
  display: none;
  flex-direction: column;

  &.active {
    display: flex;
  }

  @media screen and (min-width: 768px /*--screen-medium*/) {
    position: absolute;
    z-index: 999; //no one shall be on top of menu dropdown
    background-color: var(--menu-bg);
    border-radius: var(--border-radius-xl);

    &.active {
      padding: var(--space-sm);
    }
  }
}

.languageMenu__language {
  margin: var(--space-xxs) var(--space-sm);
  margin-bottom: 0;
}

.languageMenu__link {
  position: relative;
  color: var(--menu-text);
  text-decoration: none;

  &:hover,
  &.active:hover {
    color: var(--menu-text-hover);
  }

  &.active {
    font-weight: bold;
    color: var(--menu-text-selected);
  }
}
