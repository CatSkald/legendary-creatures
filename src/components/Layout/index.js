import "../../styles/global.scss";
import "../../styles/theme-light.scss";
import "../../styles/theme-dark.scss";
import styles from "./index.module.scss";

import React, { useState } from "react";
import Header from "../Header";
import Footer from "../Footer";

const LocaleContext = React.createContext();

const Layout = ({ children, pageContext: { language } }) => {
  const [isDarkColorTheme, switchColorTheme] = useState(false);

  function handleToggleColorTheme() {
    switchColorTheme(!isDarkColorTheme);
  }

  return (
    <LocaleContext.Provider value={{ language }}>
      <div
        className={`body ${styles.Wrapper} ${
          isDarkColorTheme ? "theme-dark" : "theme-light"
        }`}
      >
        <Header
          isDarkColorTheme={isDarkColorTheme}
          handleToggleColorTheme={handleToggleColorTheme}
        />
        <section className={styles.SiteContent} role="main">
          <div className={styles.Container}>{children}</div>
        </section>
        <Footer />
      </div>
    </LocaleContext.Provider>
  );
};

export { Layout, LocaleContext };
