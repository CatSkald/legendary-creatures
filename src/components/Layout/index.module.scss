.layout__container {
  display: flex;
  min-height: 100vh;
  flex-direction: column;
}

.layout {
  display: flex;
  flex: 1;
  width: 100%;
}

.layout__content {
  width: 100%;
  max-width: var(--container-width);
  margin: 0 auto;
  padding: var(--space-sm);

  @media screen and (min-width: 768px /*--screen-medium*/) {
    padding: var(--space);
  }
}
