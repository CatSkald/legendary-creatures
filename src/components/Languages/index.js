import "flag-icon-css/css/flag-icon.css";
import styles from "./index.module.scss";

import React from "react";
import { navigate, Link } from "gatsby";
import { LocaleContext } from "../Layout";
import usePageMapping from "../../i18n/configuration/usePageMapping";

const locales = require("../../i18n/locales");

const Languages = () => {
  const { locale } = React.useContext(LocaleContext);
  const pageMapping = usePageMapping();

  function handleClickLanguage(e, lang) {
    //TODO localStorage.getItem("preferredLanguage");
    const isBrowser = typeof window !== "undefined";
    if (isBrowser)
      window.localStorage.setItem("preferredLanguage", lang);

    if (locale === lang) {
      e.preventDefault();
      return;
    }

    const url = window.location.pathname.split("/").pop();
    if (!url) return;

    const mappedUrl = pageMapping.find(item => {
      let hasUrl = false;

      Object.entries(item).forEach(([key, value]) => {
        if (value.split("/").pop() === url) return (hasUrl = true);
      });

      return hasUrl;
    });

    if (!mappedUrl) return;

    e.preventDefault();

    return locales[lang].default
      ? navigate(`/${mappedUrl[lang]}`)
      : navigate(`/${lang}/${mappedUrl[lang]}`);
  }

  return (
    <ul className={styles.LanguageList}>
      {Object.entries(locales)
        .filter(([language, languageProps]) => !languageProps.hidden)
        .map(([language, languageProps]) => (
          <li
            className={styles.LanguageItem}
            key={languageProps.locale}
          >
            <Link
              className={styles.LanguageLink}
              key={languageProps.locale}
              to={languageProps.default ? "/" : languageProps.path}
              hrefLang={languageProps.siteLanguage}
              onClick={e =>
                handleClickLanguage(e, languageProps.siteLanguage)
              }
            >
              <span
                className={`flag-icon flag-icon-${languageProps.country}`}
              ></span>
            </Link>
          </li>
        ))}
    </ul>
  );
};

export default Languages;
