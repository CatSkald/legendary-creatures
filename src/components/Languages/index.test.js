import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import each from "jest-each";

import Languages from "../Languages";
import { LocaleContext } from "../Layout";

describe("Languages", () => {
  const createLocaleContext = language => ({
    language: {
      code: language,
    },
  });

  each([
    ["en", true],
    ["en", false],
    ["uk", true],
    ["ru", true],
  ]).it("renders correctly for %s is active %s", (language, active) => {
    const { container } = render(
      <LocaleContext.Provider value={createLocaleContext(language)}>
        <Languages
          isActive={active}
          handleLanguageSelected={jest.fn()}
          localizedLinks={[]}
        />
      </LocaleContext.Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
