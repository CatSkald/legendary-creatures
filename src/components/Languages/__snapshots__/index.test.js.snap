// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Languages renders correctly for en is active false 1`] = `
<div>
  <ul
    class="LanguageList "
  >
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink active"
        href="/"
        hreflang="en"
      >
        English
      </a>
    </li>
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink "
        href="/ru"
        hreflang="ru"
      >
        Русский
      </a>
    </li>
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink "
        href="/uk"
        hreflang="uk"
      >
        Українська
      </a>
    </li>
  </ul>
</div>
`;

exports[`Languages renders correctly for en is active true 1`] = `
<div>
  <ul
    class="LanguageList active"
  >
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink active"
        href="/"
        hreflang="en"
      >
        English
      </a>
    </li>
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink "
        href="/ru"
        hreflang="ru"
      >
        Русский
      </a>
    </li>
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink "
        href="/uk"
        hreflang="uk"
      >
        Українська
      </a>
    </li>
  </ul>
</div>
`;

exports[`Languages renders correctly for ru is active true 1`] = `
<div>
  <ul
    class="LanguageList active"
  >
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink "
        href="/"
        hreflang="en"
      >
        English
      </a>
    </li>
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink active"
        href="/ru"
        hreflang="ru"
      >
        Русский
      </a>
    </li>
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink "
        href="/uk"
        hreflang="uk"
      >
        Українська
      </a>
    </li>
  </ul>
</div>
`;

exports[`Languages renders correctly for uk is active true 1`] = `
<div>
  <ul
    class="LanguageList active"
  >
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink "
        href="/"
        hreflang="en"
      >
        English
      </a>
    </li>
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink "
        href="/ru"
        hreflang="ru"
      >
        Русский
      </a>
    </li>
    <li
      class="LanguageItem"
    >
      <a
        class="LanguageLink active"
        href="/uk"
        hreflang="uk"
      >
        Українська
      </a>
    </li>
  </ul>
</div>
`;
